Variants derive macro (#1616)
yewstack/yew: docs/concepts/components/children.md

---
title: Children
description: 
---

## General usage

_Most of the time,_ when allowing a component to have children, you don't care 
what type of children the component has. In such cases, the below example will
suffice.

```rust
use yew::prelude::*;

#[derive(Properties, Clone)]
pub struct ListProps {
	#[prop_or_default]
	pub children: Children,
}

pub struct List {
	props: ListProps,
}

impl Component for List {
	type Properties = ListProps;
	// ...

	fn veiw(&self) -> Html {
		html! {
			<div class="list">
				{ for self.props.children.iter() }
			</div>
		}
	}
}
```

## Advanced usage

Here are two cases where you would want to break away from using `Children`
and instead use something else:

 - You want to restrict the types of components that can be used as children

>    on this component.

Author: Most of this I agree with, but is objects the right term here?

 - You want to share state between a component and its children.

### Typed children

> In cases where you only need to restrict the children to a single component,

Reviewer: ```suggestion
In cases where you want one type of component to be passed as children to your component,
```

you can use `yew::html::ChildrenWithProps<T>`.

```rust
use yew::prelude::*;
use yew::html::ChildrenWithProps;

// ...

#[derive(Properties, Clone)]
pub struct ListProps {
	#[prop_or_default]
	pub children: ChildrenWithProps<Item>,
}

pub struct List {
	props: ListProps,
}

impl Component for ListProps {
	type Properties = ListProps;
	// ...

	fn view(&self) -> Html {
		html! {
			<div class="list">
				{ for self.props.children.iter() }
			</div>
		}
	}
}
```

Of course, sometimes you might need to restrict the children to a few different

> components. In these cases, you have to get a little more hands-on with yew.

Reviewer: ```suggestion
components. In these cases, you have to get a little more hands-on with Yew.
```

use yew::html::ChildrenRenderer;
use yew::virtual_dom::{ VChild, VComp };

// ...

#[derive(Clone, derive_more::From)]
pub enum ItemPropVariants {
	MyFirstComponent(MyFirstComponentProps),
	MySecondComponent(MySecondComponentProps),
}

#[derive(Clone)]
pub struct Item {
	props: ItemPropVariants,
}

impl<CHILD> From<VChild<CHILD>> for Item
where
	CHILD: Component,
	CHILD::Properties: Into<ItemPropVariants>,
{
	fn from(vchild: VChild<CHILD>) -> Self {
		Item {
			props: vchild.props.into(),
		}
	}
}

impl Into<Html> for Item {
	fn into(self) -> Html {
		match self.props {
			ItemPropVariants::MyFirstComponent(props) => {
				VComp::new::<MyFirstComponent>(props, NodeRef::default(), None).into()
			}
			ItemPropVariants::MySecondComponent(props) => {
				VComp::new::<MySecondComponent>(props, NodeRef::default(), None).into()
			}
		}
	}
}

#[derive(Properties, Clone)]
pub struct ListProps {
	#[prop_or_default]
	pub children: ChildrenRenderer<Item>,
}

pub struct List {
	props: ListProps,
}

impl Component for List {
	type Properties = ListProps;
	// ...

	fn view(&self) -> Html {
		html! {
			<div class="list">
				{ for self.props.children.iter() }
			</div>
		}
	}
}

> ```

Reviewer: I feel like a little more explanation would be good here.

Author: Will do!

Author: @teymour-aldridge I've made my first draft of an explanation in [this](https://github.com/yewstack/yew/pull/1616/commits/f02e0448a9974056d1c762bb25c65c2a94c822e3) commit. I'm relatively new-ish to Yew, so some of it might not be 100% accurate, but I think its right for the most part.

Reviewer 2: I think the explanation is great, but I think most of it could be inlined as comments. That way you get the explanation while reading the code

Author: :+1: Changed in commit [405ea75](https://github.com/yewstack/yew/pull/1616/commits/405ea755d50c4153c1c890fc9fc273456b8e4f76)

The [`derive_more`](https://github.com/JelteF/derive_more) crate is used here
for better ergonomics. If you don't want to use it, you can manually implement
`From` for each variant.

```rust
use yew::prelude::*;
use yew::html::ChildrenRenderer;
use yew::virtual_dom::{ VChild, VComp };

// `derive_more::From` implements `From<VChild<Primary>>` and
// `From<VChild<Secondary>>` for `Item` automatically!
#[derive(Clone, derive_more::From)]
pub enum Item {
    Primary(VChild<Primary>),
    Secondary(VChild<Secondary>),
}

// Now, we implment `Into<Html>` so that yew knows how to render `Item`.
impl Into<Html> for Item {
    fn into(self) -> Html {
        match self {
            Self::Primary(child) => child.into(),
            Self::Secondary(child) => child.into(),
        }
    }
}

#[derive(Properties, Clone)]
pub struct ListProps {
    #[prop_or_default]
    pub children: ChildrenRenderer<Item>,
}

pub struct List {
    props: ListProps,
}

impl Component for List {
    type Properties = ListProps;
    // ...

    fn view(&self) -> Html {
        html! {
            <div class="list">
                { for self.props.children.iter() }
            </div>
        }
    }
}
```

> ### Sharing properties

Reviewer 2: Instead of keeping you waiting even longer let me just outline what I consider to be the problem with this section and particularly the example below.

So first of all, the example kind of makes it sound like this is what you need to do to access the properties of the component when in fact it's already possible with the example from above (You can always access the properties through `VChild<T>.props` which is what's contained in `ChildrenWithProps<T>`).

A more fitting description of this section is that it's about how to limit the children to a certain type of `Component::Properties` rather than just the `Component` type from above.
In the example below any component which either has `type Properties = PrimaryProps` or `type Properties = SecondaryProps` is accepted but that brings me to the problem with it. It accepts **any** component type with either of the property types **but** it completely erases the component type itself. It always turns into a `<Primary />` or `<Secondary />` component so if we try to create our own little component which accepts these properties we would end up very confused.
I think I don't need to go into why this isn't great.

To avoid this issue `Item` would have to keep track of the original `VChild<T>` (probably in the form of `VComp` though because it isn't generic) but then we need to clone the props instead of moving them out of the `VChild` and at that point we can no longer manipulate them.
A better alternative might be to create a trait, implement it on the property types we wish to accept, and then keep a trait object `Box<dyn TheTrait>` in `Item` but that's as confusing as it probably sounds. Generally, I think there's no general solution for this. It all depends on the use case.

Unless you can think of a concrete example (or maybe a way to implement this properly) I think it's best to just remove this section for now. We can always pick it back up if it becomes a requested feature in the future.

TL;DR: let's speed things up by getting rid of this entire section, then we can merge this and move on :)

Author: Sounds good to me :+1:

---
title: Children
description: 
---

## General usage

_Most of the time,_ when allowing a component to have children, you don't care 
what type of children the component has. In such cases, the below example will
suffice.

```rust
use yew::prelude::*;

#[derive(Properties, Clone)]
pub struct ListProps {

> 	#[prop_or_default]

Author: Fixed in commit [#a0c7572](https://github.com/yewstack/yew/pull/1616/commits/a0c7572a841b2af812578369b342f12c0b58b795)

	pub children: Children,
}

pub struct List {
	props: ListProps,
}

impl Component for List {
	type Properties = ListProps;
	// ...

	fn view(&self) -> Html {
		html! {
			<div class="list">
				{ for self.props.children.iter() }
			</div>
		}
	}
}
```

## Advanced usage

There are some cases where you wouldn't want to use the `Children` type
and instead use an alternative method:

>  - You want to only allow components of a specific type to be used as children for your component.
> 
>  - You want to share state between a component and its children.

Reviewer 2: ```suggestion
 - You want to only allow components of a specific type to be used as children for your component.
 - You want to share state between a component and its children.
```

### Typed children

In cases where you want one type of component to be passed as children to your component,
you can use `yew::html::ChildrenWithProps<T>`.

```rust
use yew::prelude::*;
use yew::html::ChildrenWithProps;

// ...

#[derive(Properties, Clone)]
pub struct ListProps {
	#[prop_or_default]
	pub children: ChildrenWithProps<Item>,
}

pub struct List {
	props: ListProps,
}

impl Component for ListProps {
	type Properties = ListProps;
	// ...

	fn view(&self) -> Html {
		html! {
			<div class="list">
				{ for self.props.children.iter() }
			</div>
		}
	}
}
```

Of course, sometimes you might need to restrict the children to a few different
components. In these cases, you have to get a little more hands-on with Yew.

The `derive_more` crate is used here for better ergonomics. If you don't want

> to use it, replace `ItemPropVariants` with this:
> 
> ```rust
> #[derive(Clone)]
> pub enum ItemPropVariants {
> 	MyFirstComponent(MyFirstComponentProps),
> 	MySecondComponent(MySecondComponentProps),
> }
> 
> impl From<MyFirstComponentProps> for ItemPropVariants {
> 	fn from(props: MyFirstComponentProps) -> Self {
> 		Self::MyFirstComponent(props)		
> 	}
> }
> 
> impl From<MySecondComponentProps> for ItemPropVariants {
> 	fn from(props: MySecondComponentProps) -> Self {
> 		Self::MySecondComponent(props)
> 	}
> }
> ```

Reviewer 2: I feel like it's enough to leave it at that.
```suggestion
to use it, you can manually implement `From` for each variant.
```
Perhaps we should also link to the documentation for `derive_more::From` to encourage users to learn about what it does there?

> ```rust

Reviewer 2: One thing that needs to be mentioned for this example is that it doesn't constrain the component type but the properties type instead.
`ChildrenRenderer<Item>` accepts any component that has either `MyFirstComponentProps` or `MySecondComponentProps` as its associated properties type.
This is certainly useful if you want to manipulate the props like in the `Shared properties` section but I feel like we should also add a normal example with only restricts the components type. What do you think?

Author: I also feel like this is worth mentioning, but I'm not quite sure where to do it. Should it just be mentioned around line 82?

Reviewer 2: Yeah, that sounds like a good place to mention it.
Maybe we should modify the example to work on the component type instead of the properties? Constraining the properties type is a very specific use case, I don't think we should use it to introduce new users to the concept of typed children.
I do like the "Sharing properties" section though so how about we use it there?

Reviewer 2: What's particularly worrying about the example code right now is that it always renders the properties with either `MyFirstComponent` or `MySecondComponent`. So if we create a new component `MyOtherComponent` which also has `type Properties = MyFirstComponentProps` it won't actually use `MyOtherComponent` in the end, it will replace it `MyFirstComponent`:

```rust
html! {
  <List>
     // this will work but it will be replaced with a `MyFirstComponent` because of the `Into<Html>` implementation
    <MyOtherComponent with MyFirstComponentProps {} />
  </List>
}
```

So working with children that require a specific `Properties` type requires extra care so we don't end up in this situation.

Author: :+1: Sounds good, although I'm not sure off the top of my head how you would make it check the component type instead or properties. Might add another PR later on that adds an example for that.

Reviewer 2: That's way easier:
```rust
#[derive(Clone, derive_more::From)]
enum Item {
  First(VChild<MyFirstComponent>),
  Second(VChild<MySecondComponent>),
}
// derive_more takes care of adding `From<VChild<[First | Second]>> for Item`
impl Into<Html> for Item {
  fn into(self) -> Html {
    match self {
      Self::First(child) => child.into(),
      Self::Second(child) => child.into(),
    }
  }
}
```

That's all we need. Now `List` only accepts `MyFirstComponent` or `MySecondComponent`.

use yew::prelude::*;
use yew::html::ChildrenRenderer;
use yew::virtual_dom::{ VChild, VComp };

// ...

#[derive(Clone, derive_more::From)]
pub enum ItemPropVariants {
	MyFirstComponent(MyFirstComponentProps),
	MySecondComponent(MySecondComponentProps),
}

#[derive(Clone)]
pub struct Item {
	props: ItemPropVariants,
}

impl<CHILD> From<VChild<CHILD>> for Item
where
	CHILD: Component,
	CHILD::Properties: Into<ItemPropVariants>,
{
	fn from(vchild: VChild<CHILD>) -> Self {

> 		Item {

Reviewer 2: ```suggestion
		Self {
```

			props: vchild.props.into(),
		}
	}
}

impl Into<Html> for Item {
	fn into(self) -> Html {
		match self.props {

> 			ItemPropVariants::MyFirstComponent(props) => {
> 				VComp::new::<MyFirstComponent>(props, NodeRef::default(), None).into()
> 			}
> 			ItemPropVariants::MySecondComponent(props) => {
> 				VComp::new::<MySecondComponent>(props, NodeRef::default(), None).into()
> 			}

Reviewer 2: This should be a bit easier to understand:
```suggestion
			ItemPropVariants::MyFirstComponent(props) => {
				html! { <MyFirstComponent with props /> }
			}
			ItemPropVariants::MySecondComponent(props) => {
				html! { <MySecondComponent with props /> }
			}
```

		}
	}
}

#[derive(Properties, Clone)]
pub struct ListProps {
	#[prop_or_default]
	pub children: ChildrenRenderer<Item>,
}

pub struct List {
	props: ListProps,
}

impl Component for List {
	type Properties = ListProps;
	// ...

	fn view(&self) -> Html {
		html! {
			<div class="list">
				{ for self.props.children.iter() }
			</div>
		}
	}
}
```

Heres a run down of whats happening:

This segment here defines the different types of properties that children
components can have. Additionally, it derives `derive_more::From`, which
implements `From<MyFirstComponentProps>` and `From<MySecondComponentProps>` for
us!

```rust
#[derive(Clone, derive_more::From)]
pub enum ItemPropVariants { /* ... */ }
```

> Next, we use create this wrapper:

Reviewer 2: ```suggestion
Next, we create this wrapper:
```

```rust
#[derive(Clone)]
pub struct Item {
	props: ItemPropVariants,
}
```

Then, we implement `From<VChild<CHILD>>` for our wrapper, where `CHILD`:
 

>  - Implements `Component`
> 
>  - Has a `Self::Properties` value that implements `Into<ItemPropVariants>`.

Reviewer 2: ```suggestion
 - Implements `Component`
 - Has a `Self::Properties` value that implements `Into<ItemPropVariants>`.
```
